tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py

```diff
@@ -351,3 +351,155 @@ def forward_xp(self, inputs, xp):
         a, = inputs
         out = xp.linalg.pinv(a)
         return out,
+
+
+@op_utils.op_test(['native:0', 'cuda:0'])
+@chainer.testing.parameterize(*(
+    chainer.testing.product({
+        'shape': [(1, 1), (3, 3), (6, 6)],
+        'in_dtypes': ['float32', 'float64'],
+        'UPLO': ['u', 'L']
+    })
+))
+class TestEigh(NumpyLinalgOpTest):
+
+    def generate_inputs(self):
+        a = numpy.random.random(self.shape).astype(self.in_dtypes)
+        return a,
+
+    def forward_xp(self, inputs, xp):
+        a, = inputs
+
+        # Input has to be symmetrized for backward test to work
+        def symmetrize(A):
+            L = xp.tril(A)
+            return (L + L.T)/2.
+        a = symmetrize(a)
```

Comment:
Actually, I think gradients should be zero for the unused half. I.e. in the backward definition, if uplo is L, add the upper half to the lower half of the gradient, zeroing out the upper half. And vise versa. Then we do not need this symmetrization.

Reply:
Discussed this with the author outside this thread. I now understand that the current implementation is correct, both from a mathematical as well as a practical (as an autograd framework) standpoint. That this routine only uses either half is an "implementation detail" and should not affect the values of the gradient. It must be noted though that the routine **assumes** that the input indeed is symmetric since we choose not to perform any validation for it.